### PR TITLE
Use dedicated Document class if vector is available

### PR DESCRIPTION
**Pull request by OskarStark**
Just a first idea, was this, what you had in mind?

### CODE REVIEW

**Comment by chr-hertel** [comment]
Yes, i think inheritance, also maybe a solution for `Message` is a valid solution here

File: src/Document/VectorizedDocument.php [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+
+use Symfony\Component\Uid\Uuid;
+use Symfony\Component\Uid\UuidV4;
+
+final class VectorizedDocument extends Document
+{
+    public function __construct(
+        public Uuid $id,
+        public ?string $text,
+        public ?Vector $vector,
```
Comment (chr-hertel) [inline src/Document/VectorizedDocument.php:15]: having this nullable feels wrong with the name. and the `hasVector` could be replaced in user land code with `$document instanceof VectorizedDocument`

**Comment by chr-hertel** [comment]
I would go even one step further and change the `StoreInterface` to rely on the `VectorizedDocument`

**Comment by chr-hertel** [comment]
or is it an `EmbeddedDocument`? :thinking:

File: src/Document/VectorizedDocument.php (hunk 1 above)
Comment (OskarStark) [reply src/Document/VectorizedDocument.php:15]: Indeed 👍

**Comment by OskarStark** [comment]
> or is it an EmbeddedDocument? 🤔

Yes, why not, makes sense.

Another thing is, how we want to deal with a text? it is nullable, so what's the purpose here? should we make it required as a string in `Document`? I mean we need anyway sth. which we can use for embedding. Maybe renaming it to `$content` makes more sense?

**Comment by chr-hertel** [comment]
so the `text` is the input for the embedding. and in my understanding mostly created by (a subset) of data points of an entity or the metadata.

using your example:
```php
class Movie
{
    public string $title;
    public string $director;
    public string $description;
}
```

in simple cases we could just convert it into a text to embed like this
```
Title: {$movie->title}
Director: {$movie->director}
Description:  {$movie->description}
```
but that could a) vary on the state or meaning of the base object or b) involve other strategies like enhancing the content dynamically

**Comment by OskarStark** [comment]
Yes, I would decouple the metadata from the used string for generating the embedding.

That's why I propose to rename text to content.

In the metadata one could store what he wants, but the content string should be produced in userland without any magic IMHO

**Comment by OskarStark** [comment]
This PR is replaced by 
* https://github.com/php-llm/llm-chain/pull/95

**Comment by chr-hertel** [comment]
so we can close this and #95 and #101 belong together?

the comments are confusing me :D

**Comment by OskarStark** [comment]
Yes this was the first idea here to tackle this topic.

**Pull request closed without merging**